### DISCUSSION TITLE
Can you recommend an offer or minimum specs/connection for serving ~100 servers?

### DISCUSSION THREAD

**Discussion by ghost** [jagrosh/MusicBot#780] [Questions] [answered]
I'd like to host an instance of a music bot for a network of ~100 large (10k average users) servers.

Do you have an estimate of the minimum specs and connection I should be looking for?

**Comment by MichailiK** [comment, marked as answer]
**TL;DR**: Amount of members don't matter. It heavily relies on how many servers are playing music at once.

<hr>

The issue is not amount of members, because JMusicBot asks Discord to not send any user-related data (except messages ofc). The most concern of performance would actually be how many audio streams are going on at once: The more servers play music using your bot, the higher your resource usage becomes.

Internally, JMusicBot uses [LavaPlayer](https://github.com/sedmelluq/lavaplayer) to load & play music. There are 3 resources that JMusicBot/LavaPlayer heavily utilizes:

## 1. Memory

LavaPlayer claims it only consumes up to 350+ kilobytes per playing YouTube track. The bot itself also stores a few dozen Megabytes of data for settings, servers, channels, messages & more.

You can **partially** limit JMusicBot's ram usage [by using the `-Xmx` and `-Xms` parameters](https://stackoverflow.com/questions/14763079/what-are-the-xms-and-xmx-parameters-when-starting-jvm).

## 2. CPU

Transmitting audio can use up your CPU a bit. It's hard to be specific, but here are some considerations:
- YouTube uses Opus for audio. Discord uses Opus for audio too. This allows LavaPlayer to just mirror the audio packets to Discord, which saves CPU usage.
- Changing the volume from 100% to anything else means LavaPlayer must manipulate the audio packets, which can significantly degrade performance. This is likely why changing volume is a premium feature in some bots.
- The above point is also applicable to non-opus audio (mp3's, wav's etc.)

## 3. Networking

You will need a consistent internet connection. If your internet cuts out even by a second, users will notice the stutter. Other than that, audio files are quite small, and you won't need much bandwidth for a few audio streams.

Using [yt-dlp](https://github.com/yt-dlp/yt-dlp) to download the audio of a 2:48 minute video, i got an `.opus` audio file that was 2.6 MB large. Add to the fact that LavaPlayer is not downloading the it all at once, just like the YouTube app, and you won't need much bandwidth. 5 mbit/s is probably sufficient for a single stream, but for 10+ streams, you may need to go faster, maybe 20 mbit/s (both download & upload)? You'll have to try and see.
